core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala

```diff
@@ -441,27 +436,19 @@ trait DataSkippingReaderBase
           DataSkippingPredicate(!(min === v && max === v), minCol, maxCol)
         }
       }
-    case Not(EqualNullSafe(SkippingEligibleColumn(a), lit @ SkippingEligibleLiteral(v)))
-        if lit.value != null =>
-      val minCol = StatsColumn(MIN, a)
-      val maxCol = StatsColumn(MAX, a)
-      val nullCountCol = StatsColumn(NULL_COUNT, a)
-      // since `Not(Literal(null, _) <=> NotNullLiteral(v, _))` returns true, means we can't
-      // skipping the file which has stats with `nullCount > 0`.
-      getStatsColumnOpt(nullCountCol).flatMap { nullCount =>
-        getStatsColumnOpt(minCol).flatMap { min =>
-          getStatsColumnOpt(maxCol).map { max =>
-            DataSkippingPredicate(
-              min.isNull || max.isNull || nullCount.isNull ||
-                !(min === v && max === v && nullCount === 0),
-              minCol,
-              maxCol,
-              nullCountCol)
-          }
-        }
-      }
     case Not(EqualTo(v: Literal, a)) =>
       constructDataFilters(Not(EqualTo(a, v)))
+
+    // Match `EqualNullSafe` on a NotNullLiteral. Rewrite `EqualNullSafe(a, NotNullLiteral)` as
+    // `And(IsNotNull(a), EqualTo(a, NotNullLiteral))` to let the existing logic handle it.
+    // `EqualNullSafe(a, null)` will be replaced with `isNull(a)` by `NullPropagation` in Spark, so
+    // we don't need to handle `EqualNullSafe(a, null)`.
+    case EqualNullSafe(a, v: Literal) if v.value != null =>
```

Comment:
If `NullPropagation` is reliable, why do we need to check `v.value != null` ? If it's not reliable, should we just handle that case since it's easy to do so, instead of discarding that skipping predicate?
```scala
case EqualNullSafe(a, v: Literal) =>
  val rewrittenExpr = if (v.value != null) And(IsNotNull(a), EqualTo(a, v)) else IsNull(a)
  constructDataFilters(rewrittenExpr)
```